torch_xla/csrc/tensor_impl.cpp

```diff
@@ -178,6 +186,40 @@ void XLATensorImpl::SetupSizeProperties() {
   }
 }
 
+void XLATensorImpl::SetupSymSizeProperties() {
+  size_t generation = tensor_->generation();
+  if (generation != generation_) {
+    // Fill up the basic dimension data members which the base class
+    // implementation uses in its APIs.
+    auto shape = tensor_->shape();
+    auto rank = tensor_->shape().get().rank();
+    c10::SmallVector<c10::SymInt, 5> sym_sizes;
+    numel_ = 1;
+    for (auto i : c10::irange(rank)) {
+      // if (tensor_->shape().get().is_dynamic_dimension(i)) {
+      //   XLAIrBuilder a = XLAIrBuilder();
+      //   auto dim_node = a.MakeSizeNode(tensor_->GetIrValue(), i);
+      //   auto* sn =
+      //   dynamic_cast<torch::lazy::SymIntNodeImpl*>(dim_node.get());
+      //   sym_sizes.push_back(sn->toSymInt());
+      //   /*TODO(miladm): verify numel_ calculation after adding a dynamic op
+      //   */ numel_ *=
+      //   dynamic_cast<SizeNode*>(dim_node.get())->getStaticValue();
+      // } else {
+      sym_sizes.push_back(c10::SymInt(tensor_->shape().get().dimensions(i)));
+      numel_ *= tensor_->shape().get().dimensions(i);
+      // }
+    }
+    sizes_and_strides_.set_sizes(sym_sizes);
```

Comment:
@Krovatkin From what I can tell this pr switch to set `size` to set `sym_sizes` for tensorImpl's `sizes_and_strides_`. I am guessing upstream already support only setting sym_size?

Reply:
> From what I can tell this pr switch to set size to set sym_sizes for tensorImpl's sizes_and_strides_.

Kind of. This PR is the first step in enabling real dynamic shapes. Note, we have two separate storages here: 
* `sizes_and_strides_`
* `sym_sizes_`

Unfortunately, we can't just make C++ `sizes()` to start throwing a "NotImplemented" exception, otherwise we would break a lot of code. We want `sizes()` to return upper bounds.
This is why @miladm is populating `set_sizes_and_strides` with concrete integers wrapped in SymInts. You could think of this logic as just saving upper bounds; we just need to wrap them in SymInts since upstream is now using the unified type: SymInt.

Now the doubly unfortunate part. We know that upstream can store real symintnodes in sizes_and_strides_ wrapped in SymInts. However, we can't take advantage of that, because if we store real SymIntNodes in sizes_and_strides_ and someone calls sizes() in C++ this would trigger conversions to ints on SymIntNodes which would trigger materialization.

So we decided to have separate storage for `sym_sizes()` namely sym_sizes_.

This way when someone calls `size()` in python, we would call `XLATensor::sym_sizes_custom` which would return `sym_sizes_` which may contain both concrete ints and symintnodes. 

Now if a user uses one of those SymIntNodes in python it will trigger materialization of the SymIntNode (since we do want the exact result at least according to our discussions).

Phew....

Reply:
@JackCaoG the part I'm not 100% sure is why we need to populate `sizes_and_strides_`  in addition to `sym_sizes_` when `sym_sizes` as called. I'd think when someone calls `sizes()` that would update `sizes_and_strides_` when someone calls `sym_sizes` that would set `sym_sizes_`. 
Presumably, @miladm ran into some issues so we do need to be setting both when `sym_sizes is called.